### ISSUE TITLE
Root saga run multiple times server-side

### ISSUE THREAD

**Issue by dhovart** [bmealhouse/next-redux-saga#59]
I'm using next-redux-saga along with next-redux-wrapper. I've followed the instructions and created a `configureStore` function where I initialize my store, create the saga middleware and attach my root saga to the store with `store.sagaTask = sagaMiddleware.run(rootSaga)`.

But I've discovered that next-redux-wrapper recreates the store several times on the server. The result is that my sagas are run more than once since `sagaMiddleware.run(rootSaga)` is called each time, and I end up having multiple call to my API instead of a single one.

Is there a workaround to this issue?
Thanks in advance.

**Comment by dhovart** [comment]
It's basically the same issue as this one: https://github.com/kirill-konshin/next-redux-wrapper/issues/122
But there is no satisfying answers.

**Comment by bbortt** [comment]
Hi @dhovart 
Can you provide some code, please?

**Comment by dhovart** [comment]
Sure thing.

store/init.ts

```ts
import { createStore, applyMiddleware, Store, Middleware, StoreEnhancer } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware, { Task } from 'redux-saga';
import thunkMiddleware from 'redux-thunk';
import { rootReducer, AppState } from './reducers';
import { rootSaga } from './sagas';

export interface StoreWithSaga extends Store<AppState> {
    sagaTask: Task;
}

const bindMiddleware = (middleware: Middleware[]): StoreEnhancer => {
    if (process.env.NODE_ENV !== 'production') {
        return composeWithDevTools(applyMiddleware(...middleware));
    }
    return applyMiddleware(...middleware);
};

export const initStore = (preloadedState?: AppState): StoreWithSaga => {
    const sagaMiddleware = createSagaMiddleware();

    const store = createStore(
        rootReducer,
        preloadedState,
        bindMiddleware([thunkMiddleware, sagaMiddleware])
    ) as StoreWithSaga;

    store.sagaTask = sagaMiddleware.run(rootSaga);

    return store;
};
```

store/sagas.ts

```ts
import { SagaIterator } from 'redux-saga';
import { all, call } from 'redux-saga/effects';
import { homepageSaga } from '../modules/homepage';
import { authSaga } from '../modules/auth';

export function* rootSaga(): SagaIterator {
    yield all([
        call(homepageSaga),
        call(authSaga),
    ]);
}
```

store/reducers.ts

```ts
import { combineReducers } from 'redux';
import { homepageReducer, HomepageState } from '../modules/homepage';
import { authReducer, AuthState } from '../modules/auth';

export interface AppState {
    homepage: HomepageState;
    auth: AuthState;
}

export const rootReducer = combineReducers<AppState>({
    homepage: homepageReducer,
    auth: authReducer,
});
```


pages/_app.tsx

```ts
import React from 'react';
import App, { Container } from 'next/app';
import { appWithTranslation } from '../i18n';
import { Provider } from 'react-redux';
import withRedux from 'next-redux-wrapper';
import withReduxSaga from 'next-redux-saga';
import { initStore } from '../store/init';
import { NextComponentType, NextContext } from 'next';
import { Store, compose } from 'redux';
import { AppState } from '../store/reducers';

interface BaseAppProps {
    pageProps: object;
}

type BaseAppPropsWithStore = BaseAppProps & { store: Store<AppState> };

class MyApp extends App<BaseAppPropsWithStore> {
    public static async getInitialProps({
        Component,
        ctx,
    }: {
        Component: NextComponentType;
        ctx: NextContext;
    }): Promise<BaseAppProps> {
        return {
            pageProps: Component.getInitialProps ? await Component.getInitialProps(ctx) : {},
        };
    }

    public render(): JSX.Element {
        const { Component, pageProps, store } = this.props;
        return (
            <Container>
                <Provider store={store}>
                    <Component {...pageProps} />
                </Provider>
            </Container>
        );
    }
}

export default compose(
    withRedux(initStore),
    appWithTranslation,
    withReduxSaga
)(MyApp);
```

And an example of a saga called several times, AuthSaga.ts :

```ts
import { SagaIterator } from 'redux-saga';
import { all, call, take, fork, put } from 'redux-saga/effects';
import { AuthActionType } from './AuthActionType';
import axios, { AxiosPromise, AxiosResponse } from 'axios';
import APIConstants from '../../constants/API';
import { authenticationError, authenticationSuccess, authenticatedTokenRequest, Credentials } from './AuthActions';

function getOauthState(): AxiosPromise<AxiosResponse> {
    return axios.get(`${APIConstants.AUTH_ENDPOINT}/oauth-state`, { withCredentials: true });
}

function* checkAuthentication(): SagaIterator {
    const oauthStateResponse = yield call(getOauthState);

    if (oauthStateResponse.data.authenticated === true) {
        yield put(authenticationSuccess());
        yield put(authenticatedTokenRequest());
    }
    return;
}

// ...

export function* authSaga(): SagaIterator {
    yield all([call(checkAuthentication), call(watchAuthenticationRequest), call(watchDeauthenticationRequest)]);
}
```

or HomepageSaga.ts:

```ts
import { takeLatest, call } from 'redux-saga/effects';
import * as actions from './HomepageActions';
import { HomepageActionType } from './HomepageActionType';
import { fetchApi } from '../../service';
import { SagaIterator } from 'redux-saga';
import { AxiosResponse } from 'axios';

function* fetchHomepage(): SagaIterator {
    yield call(fetchApi, {endpoint: 'homepage' });
}

export function* homepageSaga(): SagaIterator {
    yield takeLatest(HomepageActionType.FETCH_HOMEPAGE, fetchHomepage);
}
```

I've not included everything, but whenever I log the axios calls, I can see that there are several calls to the homepage endpoint and the oauth-state endpoint. The number of time this appears in the logs varies randomly.

**Comment by bbortt** [comment]
thx. i'll have a look at it shortly 🚀

**Comment by dhovart** [comment]
@bbortt Any news on this ? Thank you

**Comment by bbortt** [comment]
Sorry @dhovart I was busy and totally forgot about this.
I remember I looked into the snippets above and thought it looks good, except for I don't know what the impact of `compose` is. I usually write it like this:
```
export default withRedux(configureStore)(
  withReduxSaga(ReduxContextAwareAppClass)
)
```
I think I've got a working example [here](https://github.com/bbortt/qdrakeboo/tree/master/source/frontend/next-js-web-ui): You can check my code.

**Comment by dhovart** [comment]
Yes, I don't see anything wrong with this code either… But the issue is that my sagas are rerun several times on the server.

I'll try to put together a small repo tonight so that you can see for yourself.

`compose` is just a small utility that composes functions together, it shouldn't have any impact.
https://github.com/reduxjs/redux/blob/master/src/compose.ts#L101 
It boils down to what you're already doing in your example, except that I find it more convenient to use and more legible when there are more than 2 HOCs to be composed together.

**Comment by dhovart** [comment]
I've created a repository here: https://github.com/dhovart/next-redux-saga-multiple-calls

You can see my root saga is called multiple times at startup (I guess every time the store is recreated by next-redux-wrapper). On the current project I'm working on, we're logging the requests, and sometimes it's just spamming the terminal (there can be 10+ in a row).

**Comment by bbortt** [comment]
what I've investigated so far:
```
=> first handle (request to /)
app: init store
app: running root saga
app: fetching auth state…
next-redux-saga: getInitialProps - server? -> true
app: fetching homepage...
next-redux-saga: stopping saga.. (server side)
app: User is not authenticated
next-redux-saga: done.
app: init store
app: running root saga
app: fetching auth state…
app: User is not authenticated
=> request done, response sent
=> second handle
```
this seems to be an issue with either `next-redux-wrapper` or I think rather `next.js` itself. because the second initialization of the store does not even trigger `next-redux-saga` anymore.
requesting help from @bmealhouse at this point.

**Comment by bbortt** [comment]
So in my application it looks like this:
```js
init store
running root saga
next-redux-saga#getInitialProps - server side?  true
_app#getInitialProps - server side?  true
2019-10-04T08:05:24.794Z [Next.js Webapplication] [INFO] Proxying request to 'http://localhost:8080/userinfo'
stopping saga..
dispatching @@redux-saga/CHANNEL_END
next state: {"alert":{"alerts":[]},"health":{"online":true},"userInfo":{"isAuthenticated":false,"userInfo":{},"permissions":[]},"userManagement":{"resetPassword":{"loading":false,"error":""}}}
done.
init store
running root saga
```
see, there is no `_app#getInitialProps` the second time. no chance for us to catch this. the problem is not located in our hook, I'm 100% sure.

**Comment by dhovart** [comment]
Thank you for investigating !
I've tried (without much hope) storing the initial task and reattaching it later on to the store but it's not working. 
See: https://github.com/dhovart/next-redux-saga-multiple-calls/commit/8bc2228ebe209f8ca2c62f94282dcec426f6253e

I guess it's because it's dispatching action to the previous store or that it loses its reference to it…

**Comment by bbortt** [comment]
The saga (it's a channel) is destroyed by dispatching `END`. You won't have any chance to restore it ;)

**Comment by dhovart** [comment]
Ah right, I feel stupid saying this but I hadn't really looked at how the HOC worked.

**Comment by bbortt** [comment]
@kirill-konshin helped me out with this:
> Due to changes in 2.1.0 store should be recreated multiple times. Once clean and once with pre-hydrated data to emulate client-side behavior. I cannot help you with sagas because I don't use them at all...

the initial state when creating the store in `const configureStore = (state = initialState) => {` is `state === undefined || initialState`. then it will be `state === initialState || whateverTheStateWasModifiedTo` in the second initialization.

@bmealhouse we should probably add and check a property in the state (something like `state.saga.initialized = true`). I'll think about it and probably propose a pull-request tomorrow.

**Comment by bbortt** [comment]
Closing this as `won't fix`.
We (as a HOC) can neither access nor modify the current application state. The user must take care of a good implementation by his own. I've updated the documentation. An example implementation is available in [`bbortt/qdrakeboo:8e39ae13274768d6e6a36a883efb05d01bb67b1d`](https://github.com/bbortt/qdrakeboo/commit/8e39ae13274768d6e6a36a883efb05d01bb67b1d).
